numba/np/ufunc/dufunc.py

```diff
@@ -778,6 +779,136 @@ def impl_axis_none(ufunc,
+                raise errors.NumbaTypeError(msg)
+
+            if not numpy_support.type_can_asarray(indices):
+                msg = 'The second argument "indices" must be array-like'
+                raise errors.NumbaTypeError(msg)
+
+            if not isinstance(axis, (types.Integer, types.Omitted,
+                                     types.IntegerLiteral, int)):
+                msg = '"axis" must be an integer'
+                raise errors.NumbaTypeError(msg)
+
+            out_none = cgutils.is_nonelike(out)
+            if not out_none and not isinstance(out, types.Array):
+                raise errors.NumbaTypeError('output must be an array')
+
+            array_arr = isinstance(array, types.Array)
+            array_ndim = array.ndim if array_arr else 0
+            tup_m1 = (0,) * (array_ndim - 1)
+            indices_arr = isinstance(indices, types.Array)
+            dt = array.dtype if cgutils.is_nonelike(dtype) else dtype
+
+            need_cast = not (array_arr and indices_arr)
+
+            if indices_arr and indices.ndim != 1:
+                msg = ('Expect "indices" array to have at most 1 dimension. '
+                       f'Got {indices.ndim}')
+                raise errors.NumbaValueError(msg)
+
+            if need_cast:
+                def impl_cast(ufunc, array, indices, axis=0, dtype=None, out=None):  # noqa: E501
+                    return ufunc.reduceat(np.asarray(array),
+                                          np.asarray(indices),
+                                          axis,
+                                          dtype=dtype,
+                                          out=out)
+                return impl_cast
+
+            @register_jitable
+            def tuple_slice(tup, pos):
+                s = tup_m1  # TODO: replace this by a literally call
+                i, j = 0, 0
+                while i < len(tup):
+                    if i != pos:
+                        s = tuple_setitem(s, j, tup[i])
+                        j += 1
+                    i += 1
+                return s
+
+            # importing it here as an import at the top scope brings unwanted
+            # stuff. See numba/tests/test_import.py::test_no_impl_import
+            from numba.np.arrayobj import generate_getitem_setitem_with_axis
+            setitem = generate_getitem_setitem_with_axis(array.ndim, 'setitem')
+
+            def impl(ufunc, array, indices, axis=0, dtype=None, out=None):
+                sz = indices.shape[0]
+
+                ax = axis
+                if axis < 0:
+                    axis += array_ndim
+
+                if axis < 0 or axis >= array_ndim:
+                    raise ValueError(f"axis {ax} is out of bounds for array "
+                                     f"of dimension {array_ndim}")
+
+                shape = tuple_setitem(array.shape, axis, sz)
+
+                if not out_none and out.shape != shape:
+                    # TODO: improve error message
```

Comment:
Does this TODO needs to be addressed? (If not, can it be removed)

Reply:
I could include the shapes in the error message, but this would require https://github.com/numba/numba/pull/9524 to be merged first. Would it be ok to update the TODO message to include a link to the PR as well. And I can update the exception message once (if) the PR gets merged.